File: design-patterns/go/builder/vehicle/plane_builder.go
package main

import "fmt"

// plane_builder.go

type PlaneBuilder struct {
	wheel    int
	engine   int
	seat     int
	interior bool
	door     int
	wing     int
}

func NewPlaneBuilder() (planeBuilder *PlaneBuilder) {
	planeBuilder = &PlaneBuilder{}
	return
}

func (planeBuilder *PlaneBuilder) AddDoor(noOfDoor int) {
	fmt.Printf("Add %d door\n", noOfDoor)
	planeBuilder.door += noOfDoor
}

func (planeBuilder *PlaneBuilder) AddEngine(noOfEngine int) {
	fmt.Printf("Add %d engine\n", noOfEngine)
	planeBuilder.engine += noOfEngine
}

func (planeBuilder *PlaneBuilder) AddInterior() {
	fmt.Println("Add interior")
	planeBuilder.interior = true
}

func (planeBuilder *PlaneBuilder) AddSeat(noOfSeat int) {
	fmt.Printf("Add %d Seat\n", noOfSeat)
	planeBuilder.seat = noOfSeat
}

func (planeBuilder *PlaneBuilder) AddWheel(noOfWheel int) {
	fmt.Printf("Add %d wheels\n", noOfWheel)
	planeBuilder.wheel += noOfWheel
}

func (planeBuilder *PlaneBuilder) AddWing(noOfWings int) {
	fmt.Printf("Add %d wing\n", noOfWings)
	planeBuilder.wing += planeBuilder.wing
}

func (planeBuilder *PlaneBuilder) Build() *Plane {
	plane := NewPlane(planeBuilder.wheel, planeBuilder.engine, planeBuilder.seat, planeBuilder.door, planeBuilder.wing, planeBuilder.interior)
	return plane
}

File: design-patterns/ts/builder/vehicle/plane-builder.ts
// plane-builder.ts

import Plane from "./plane";
import VehicleBuilder from "./vehicle-builder";

class PlaneBuilder implements VehicleBuilder{
    private wheel: number = 0;
    private engine: number = 0;
    private seat!: number;
    private interior!: boolean;
    private door: number = 0;
    private wing: number = 0;

    addWheel(noOfWheel: number) {
        console.log("Add " + noOfWheel + " wheels");
        this.wheel += noOfWheel;
    }

    addEngine(noOfEngine: number) {
        console.log("Add " + noOfEngine + " engine");
        this.engine += noOfEngine;
    }

    addSeat(noOfSeat: number) {
        console.log("Add " + noOfSeat + " Seat");
        this.seat = noOfSeat;
    }

    addInterior() {
        console.log("Add interior");
        this.interior = true;
    }

    addDoor(noOfDoor: number) {
        console.log("Add " + noOfDoor + " door");
        this.door += noOfDoor;
    }

    addWing(noOfWings: number) {
        console.log("Add " + noOfWings + " wing");
        this.wing += noOfWings;
    }

    build(): Plane {
        const plane: Plane = new Plane(this.wheel, this.engine, this.seat, this.door, this.wing, this.interior);

        return plane;
    }
}

export default PlaneBuilder;